quinn/src/connection.rs

```diff
@@ -310,6 +310,26 @@ impl Connection {
         self.0.lock().unwrap().inner.server_name().map(|x| x.into())
     }
 
+    /// The certificate chain used by the peer to authenticate
+    ///
+    /// For clients, this is the certificate chain of the server. For servers,
+    /// it is the certificate chain of the client, or [`None`] if client
+    /// authentication was not requested.
```

Comment:
Is this always true? Are clients unable to volunteer certificates unprompted?

Reply:
> Are clients unable to volunteer certificates unprompted?

Based on some skimming of the specs, that does appear to be true. There's a nice graphic here: https://comodosslstore.com/blog/what-is-ssl-tls-client-authentication-how-does-it-work.html.

Reply:
My understanding of [Section 4.6.2](https://tools.ietf.org/html/rfc8446#section-4.6.2) of the TLS 1.3 RFC is that the client indicates that it is _willing_ to provide client authentication post-handshake. The server then sends `CertificateRequest` to make it actually present the cert.

Reply:
(And if that fails, we shouldn't get here)

Reply:
That documents the procedure for a server to request client authentication. What happens if the client provides it without a request?

Reply:
Ya good point. `rustls`‘ state machine rejects invalid transitions:

https://github.com/ctz/rustls/blob/master/rustls/src/server/tls13.rs#L606
https://github.com/ctz/rustls/blob/master/rustls/src/server/tls13.rs#L641
https://github.com/ctz/rustls/blob/master/rustls/src/handshake.rs#L14